keyboards/handwired/owlet60/matrix.c

```diff
@@ -0,0 +1,323 @@
+{
+    uint8_t count = 0;
+    for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
+        count += matrix_bitpop(i);
+    }
+    return count;
+}
+
+// uses standard row code
+static void select_row(uint8_t row)
+{
+    setPinOutput(row_pins[row]);
+    writePinLow(row_pins[row]);
+}
+
+static void unselect_row(uint8_t row)
+{
+    setPinInputHigh(row_pins[row]);
+}
+
+static void unselect_rows(void)
+{
+    for(uint8_t x = 0; x < MATRIX_ROWS; x++) {
+        setPinInputHigh(row_pins[x]);
+    }
+}
+
+static void init_pins(void) {   // still need some fixing, this might not work
+  unselect_rows();              // with the loop
+  /*
+  for (uint8_t x = 0; x < MATRIX_COLS; x++) {
+    setPinInputHigh(col_pins[x]);
+  }
+  */
+  setPinInputHigh(dat_pin);
+}
+
+static bool read_cols_on_row(matrix_row_t current_matrix[], uint8_t current_row)
+{
+    // Store last value of row prior to reading
+    matrix_row_t last_row_value = current_matrix[current_row];
+
+    // Clear data in matrix row
+    current_matrix[current_row] = 0;
+
+    // Select row and wait for row selecton to stabilize
+    select_row(current_row);
+    wait_us(30);
+
+    // For each col...
+    for(uint8_t col_index = 0; col_index < MATRIX_COLS; col_index++) {
+
+        // Select the col pin to read (active low)
+        select_col_manual(col_index);
+        wait_us(30);
+        uint8_t pin_state = readPin(dat_pin);
+
+        // Populate the matrix row with the state of the col pin
+        current_matrix[current_row] |=  pin_state ? 0 : (ROW_SHIFTER << col_index);
+    }
+
+    // Unselect row
+    unselect_row(current_row);
+
+    return (last_row_value != current_matrix[current_row]);
+}
+
+
+void matrix_init(void) {
+
+    // initialize key pins
+    init_pins();
+
+    // initialize matrix state: all keys off
+    for (uint8_t i=0; i < MATRIX_ROWS; i++) {
+        raw_matrix[i] = 0;
+        matrix[i] = 0;
+    }
+
+    debounce_init(MATRIX_ROWS);
+
+    matrix_init_quantum();
+
+    setPinInput(D5);
+   writePinLow(D5);
+
+   setPinInput(B0);
+   writePinLow(B0);
+}
+
+// modified for per col read matrix scan
+uint8_t matrix_scan(void)
+{
+    bool changed = false;
+        
+    for (uint8_t current_row = 0; current_row < MATRIX_ROWS; current_row++) {
+        changed |= read_cols_on_row(raw_matrix, current_row);
+    } 
+
+    debounce(raw_matrix, matrix, MATRIX_ROWS, changed);
+
+    matrix_scan_quantum();
+    return (uint8_t)changed;
+}
+
+/*
+uint8_t matrix_scan(void)
+{
+  bool changed = false;
+
+#if (DIODE_DIRECTION == COL2ROW)
+  // Set row, read cols
+  for (uint8_t current_row = 0; current_row < MATRIX_ROWS; current_row++) {
+    changed |= read_cols_on_row(raw_matrix, current_row);
+  }
+#endif
+
+  debounce(raw_matrix, matrix, MATRIX_ROWS, changed);
+
+  matrix_scan_quantum();
+  return (uint8_t)changed;
+}
+*/
+
+static void select_col_manual(uint8_t col) {
+    // emergency code, just a simple switch statement
+    
+    if(col == 0) 
```

Comment:
for readability I would use a switch instead of if then else. It would also probably be more readable to just use an 2d array of 0, 1; set pin to low for 0 and set pin to high for 1; there are only 3 columns so, that's a 3 x 8 array, which isn't too bad. I think it would end up a lot more readable and less repetitive.

Reply:
Hello yanfali. Thank you for your suggestion. I've implemented a new select code with your suggestion in mind. If you think there's anything that I could improve upon, please tell me.